browser/static/js/main_zstack.js

```diff
@@ -1036,68 +960,86 @@ function handleMousemove(evt) {
+    if (Object.prototype.hasOwnProperty.call(payload.imgs, 'seg_arr')) {
+      canvas.segArray = payload.imgs.seg_arr;
+    }
 
-        if (Object.prototype.hasOwnProperty.call(payload.imgs, 'segmented')) {
-          adjuster.segLoaded = false;
-          adjuster.segImage.src = payload.imgs.segmented;
-        }
+    if (Object.prototype.hasOwnProperty.call(payload.imgs, 'segmented')) {
+      adjuster.segLoaded = false;
+      adjuster.segImage.src = payload.imgs.segmented;
+    }
 
-        if (Object.prototype.hasOwnProperty.call(payload.imgs, 'raw')) {
-          adjuster.rawLoaded = false;
-          adjuster.rawImage.src = payload.imgs.raw;
-        }
-      }
-      if (payload.tracks) {
-        tracks = payload.tracks;
-        // update maxLabelsMap when we get new track info
-        for (let i = 0; i < Object.keys(tracks).length; i++) {
-          const key = Object.keys(tracks)[i]; // the keys are strings
-          if (Object.keys(tracks[key]).length > 0) {
-            // use i as key in this map because it is an int, mode.feature is also int
-            maxLabelsMap.set(i, Math.max(...Object.keys(tracks[key]).map(Number)));
-          } else {
-            // if no labels in feature, explicitly set max label to 0
-            maxLabelsMap.set(i, 0);
-          }
-        }
-      }
-      if (payload.tracks || payload.imgs) {
-        render_image_display();
+    if (Object.prototype.hasOwnProperty.call(payload.imgs, 'raw')) {
+      adjuster.rawLoaded = false;
+      adjuster.rawImage.src = payload.imgs.raw;
+    }
+  }
+  if (payload.tracks) {
+    tracks = payload.tracks;
+    // update maxLabelsMap when we get new track info
+    for (let i = 0; i < Object.keys(tracks).length; i++) {
+      const key = Object.keys(tracks)[i]; // the keys are strings
+      if (Object.keys(tracks[key]).length > 0) {
+        // use i as key in this map because it is an int, mode.feature is also int
+        maxLabelsMap.set(i, Math.max(...Object.keys(tracks[key]).map(Number)));
+      } else {
+        // if no labels in feature, explicitly set max label to 0
+        maxLabelsMap.set(i, 0);
       }
-    },
-    async: false
-  });
+    }
+  }
+  if (payload.tracks || payload.imgs) {
+    render_image_display();
+  }
+}
+
+function action(action, info) {
+  backendAction = new BackendAction(action, info);
+  actions.addFencedAction(backendAction);
+}
+
+function undo() {
+  actions.undo();
+  mode.clear();
+  render_image_display();
+}
+
+function redo() {
+  actions.redo();
+  mode.clear();
+  render_image_display();
+}
+
+function displayUndoRedo() {
+  let canvasElement = document.getElementById('canvas');
+  let undoButton = document.getElementById('undo');
+  undoButton.hidden = false;
+  undoButton.style.width = canvasElement.width / 2 + 'px';
+  
+  let redoButton = document.getElementById('redo');
+  redoButton.hidden = false;
+  redoButton.style.width = canvasElement.width / 2 + 'px';
```

Comment:
I put both of these (setting hidden to false & setting the width) on the javascript side because they are dynamic behaviors. We hide the undo/redo buttons until the image is loaded, and we adjust the width of the buttons ti match the canvas width whenever we resize the window as the canvas resized by
```
    // resize the canvas every time the window is resized
    window.addEventListener('resize', function() {
      waitForFinalEvent(() => {
        mode.clear();
        setCanvasDimensions(payload.dimensions);
        brush.refreshView();
        displayUndoRedo(); // match the button width to the canvas width
      }, 500, 'canvasResize');
    });
```

Is there a way to dynamically adjust these attributes just with CSS/Materialize classes?

Reply:
Yes, materialize can dynamically size elements according to their column sizes. However, I notice now that we are changing the size of the canvas with JS and linking the buttons to that canvas size. This is not nearly as easy to resolve with CSS, so let's not worry about it.